ui/src/components/Header/Controller.tsx

```diff
@@ -32,49 +38,110 @@ export const Controller = (): ReactElement => {
-    } else {
-      ddClient.desktopUI.toast.success('Starting LocalStack');
-    }
-    const corsArg = '-e EXTRA_CORS_ALLOWED_ORIGINS=http://localhost:3000';
+  const normalizeArguments = async () => {
+    const extendedFlag = FLAGS;
+
+    const corsArg = ['-e', `EXTRA_CORS_ALLOWED_ORIGINS=${CORS_ALLOW_DEFAULT}`];
     const addedArgs = runConfig.find(config => config.name === runningConfig)
       .vars.map(item => {
-        if (item.variable === 'EXTRA_CORS_ALLOWED_ORIGINS') {
+        if (item.variable === 'EXTRA_CORS_ALLOWED_ORIGINS') { // prevent overriding variable
           corsArg.slice(0, 0);
-          return ['-e', `${item.variable}=${item.value},http://localhost:3000`];
+          return ['-e', `${item.variable}=${item.value},${CORS_ALLOW_DEFAULT}`];
+        }
+        if (item.variable === 'DOCKER_FLAGS') {
+          extendedFlag[1] = FLAGS.at(1).slice(0, -1).concat(` ${item.value}'`);
         }
+
         return ['-e', `${item.variable}=${item.value}`];
       }).flat();
 
-
     const standardDir = `${ddClient.host.platform === 'darwin' ? 'Users' : 'home'}/${mountPoint}`;
-    const mountArg = `-e LOCALSTACK_VOLUME_DIR=/${mountPoint === 'tmp' ? mountPoint : standardDir}/.localstack-volume`;
-      
-    ddClient.docker.cli.exec('run', [mountArg, corsArg, ...addedArgs, ...START_ARGS]).then(() => mutate());
+    const mountArg = ['-e', `LOCALSTACK_VOLUME_DIR=/${mountPoint === 'tmp' ? `${mountPoint}` :
+      `${standardDir}/.cache`}/localstack/volume`];
+
+    return [...extendedFlag, ...mountArg, ...corsArg, ...addedArgs, ...START_ARGS];
+  };
+
+  const start = async () => {
+    const images = await ddClient.docker.listImages() as [DockerImage];
+    const haveLocally = images.some(image => image.RepoTags?.at(0) === LATEST_IMAGE);
+
+    if (!haveLocally) {
+      setDownloadProps({ open: true, image: LATEST_IMAGE });
+      return;
+    }
+    const args = await normalizeArguments();
+
+    setIsStarting(true);
+    ddClient.docker.cli.exec('run', args, {
+      stream: {
+        onOutput(data): void {
+          if (data.stderr && !data.stderr.includes('Successfully') && // Api key activation is included in the error stream
+            !data.stderr.includes('Execution of "prepare_host"')) { 
+            ddClient.desktopUI.toast.error(data.stderr);
+            setIsStarting(false);
+          }
+        },
+        onClose(exitCode) {
+          setIsStarting(false);
+          if (exitCode === 0) {
+            ddClient.desktopUI.toast.success('LocalStack started');
+          }
+        },
+      },
+    });
   };
 
   const stop = async () => {
-    ddClient.docker.cli.exec('run', STOP_ARGS).then(() => mutate());
+    setIsStopping(true);
+    const containers = await ddClient.docker.listContainers({ 'all': true }) as [DockerContainer];
+
+    const stoppedContainer = containers.find(container =>
+      (container.Image === 'localstack/localstack' ||
+        container.Image === 'localstack/localstack-pro') &&
```

Comment:
consistency is key, please either have all operators at the end of the line, or at the start,
I personally prefer at the start.